### ISSUE TITLE
A way to mark preprocessor macros as deprecated

### ISSUE THREAD

**Issue by CelticMinstrel** [wesnoth/wesnoth#1866]
My idea is to add a new preprocessor directive that can be used in a macro declaration to mark it deprecated.

**Comment by gfgtdf** [comment]
I like this idea but i wonder: even without this, couldn't we jut put a #wanring insie the #define that is deprecated for example:
```
#define ABILITY_LEADERSHIP_LEVEL_4
#warning ABILITY_LEADERSHIP_LEVEL_4 is deprecated
    {ABILITY_LEADERSHIP}
#enddef
```

**Comment by CelticMinstrel** [comment]
I'm not sure. Does that work? Does it print the message only when the macro is actually used? If so, I suppose we could close this (or maybe even just define a macro that interpolates some values into a predefined message?).

**Comment by Vultraz** [comment]
A cursory test seems to indicate that is indeed the case. Needs more testing though.

**Comment by CelticMinstrel** [comment]
Even if it works though, it may be nice to have a separate `#deprecated` which automatically formats a stock message from its arguments. That would also allow us to translate deprecation messages.

**Comment by gfgtdf** [comment]
For deprecated i think it'd also be nice if it was displayed only at most once (for each addon),

**Comment by CelticMinstrel** [comment]
Actually, in my opinion that's _not_ a good idea. If it displays for every use of the macro, then that means you have a nice convenient list of everything that needs fixing.

Admittedly, you could always just use a search function to find all the locations once you know the specific macro that needs to be replaced, but I still think it's more convenient to have an exact list of locations.

**Comment by gfgtdf** [comment]
I still vote for using a seach function since it's usually easier to use. For exampel in my editrs seach function i can easily just click on that match to open that line in my editor. Also this is more a problem when you use that macro inside another macro, lets have `#define A {DEPRECATED_CORE_MACRO}` and you use `{A} a lot in your code you get spammed by useless deprecation wanrings`even though you just need to fix one part of you core.

**Comment by CelticMinstrel** [comment]
Search probably isn't easier to use for many Windows users, and may not be for some Mac users either. This is because it's not a single-file search - you need to search through every cfg file in your addon.

In the case of a macro within a macro, ideally the deprecation message would print only once when the outer macro is first defined. I don't know how easy this is to do within the preprocessor, though, so you may have a point there.
